### ISSUE TITLE
Non-upper case constant warning misapplied in patterns

### ISSUE THREAD

**Issue by rpjohnst** [rust-lang/rust#39371]
A constant with `#[allow(non_upper_case_globals)]` applied still generates a warning when used in a pattern, unless the `match` also has the allow attribute. For example:

```rust
#[allow(non_upper_case_globals)]
mod keywords {
    pub const If: u32 = 0;
    pub const While: u32 = 1;
}

fn main() {
    use keywords::*;

    // the warning goes away with this attribute:
    // #[allow(non_upper_case_globals)]
    match 0 {
        keywords::If => println!("if"),
        While => println!("while"),
        _ => println!("<unknown>"),
    }
}
```

[Playground](https://is.gd/jzNZQB)

**Comment by est31** [comment]
The code that emits that warning links to issue #7526. One should find out what its actual purpose was and whether its still relevant today.

**Comment by mgsloan** [comment]
Is it sensible to have this warning at all in usages of the global?  After all, the user of a library isn't really free to make this choice.  How about just when the global / constant is defined?  I'm encountering this for all of the constants from xlib - https://docs.rs/x11/2.15.0/x11/xlib/index.html#constants .  For ease of copy+modifying (or [corroding](https://github.com/jameysharp/corrode)!) C code, this choice makes sense.

Note: I'm rather new to rust.  So, perhaps there's a good reason for this that I don't see.  From the newbie perspective this is a puzzling warning.

**Comment by myfreeweb** [comment]
haha, I'm deriving a trait for an enum that uses bindgen-generated constants:

```
warning: constant in pattern `weston_layer_position_WESTON_LAYER_POSITION_UI` should have an upper case name such as `WESTON_LAYER_POSITION_WESTON_LAYER_POSITION_UI`        
  --> src/layer.rs:22:41                                                                                                                                                                
   |                                                                                                                                                                                    
22 | #[derive(Debug, Copy, Clone, PartialEq, Primitive)]                                                                                                                                
   |                                         ^^^^^^^^^                                                                                                                                  
```

I can't even put `#[allow(non_upper_case_globals)]` anywhere in my code :D

**Comment by Azorlogh** [comment]
When is this warning ever even needed?
